python/cudf/cudf/core/column/categorical.py

```diff
@@ -191,108 +206,135 @@ def _set_categories(self, new_categories, **kwargs):
+        header = {}
+        frames = []
         header["type"] = pickle.dumps(type(self))
-        header["dtype"] = self._dtype.str
-        frames.extend(category_frames)
+        header["dtype"], dtype_frames = self.dtype.serialize()
+        header["dtype_frames_count"] = len(dtype_frames)
+        frames.extend(dtype_frames)
+        header["data"], data_frames = self.codes.serialize()
+        header["data_frames_count"] = len(data_frames)
+        frames.extend(data_frames)
+        if self.mask is not None:
+            mask_frames = [self._mask_view()]
+        else:
+            mask_frames = []
+        frames.extend(mask_frames)
         header["frame_count"] = len(frames)
         return header, frames
 
     @classmethod
     def deserialize(cls, header, frames):
-        data, mask = super(CategoricalColumn, cls).deserialize(header, frames)
-
-        # Handle categories that were serialized as a cudf.Column
-        category_frames = frames[
-            len(frames) - header["category_frame_count"] :
-        ]
-        cat_typ = pickle.loads(header["categories"]["type"])
-        _categories = cat_typ.deserialize(
-            header["categories"], category_frames
+        n_dtype_frames = header["dtype_frames_count"]
+        dtype = CategoricalDtype.deserialize(
+            header["dtype"], frames[:n_dtype_frames]
         )
+        n_data_frames = header["data_frames_count"]
 
-        categories = column.as_column(_categories)
-
-        return cls(
-            data=data,
-            mask=mask,
-            categories=categories,
-            ordered=header["ordered"],
+        column_type = pickle.loads(header["data"]["type"])
+        data = column_type.deserialize(
+            header["data"],
+            frames[n_dtype_frames : n_dtype_frames + n_data_frames],
+        )
+        mask = None
+        if header["frame_count"] > n_dtype_frames + n_data_frames:
+            mask = Buffer(frames[n_dtype_frames + n_data_frames])
+        return column.build_column(
+            data=None, dtype=dtype, mask=mask, children=(data,)
         )
-
-    def _replace_defaults(self):
-        params = super(CategoricalColumn, self)._replace_defaults()
-        params.update(categories=self._categories, ordered=self._ordered)
-        return params
 
     @property
     def as_numerical(self):
-        from cudf.core.column import numerical
-
-        return self.view(numerical.NumericalColumn, dtype=self.data.dtype)
+        return column.build_column(
+            data=self.codes.data,
+            dtype=self.codes.dtype,
+            mask=self.mask,
+            name=self.name,
+        )
 
     @property
     def categories(self):
-        return self._categories
+        return self.dtype.categories.as_column()
+
+    @categories.setter
+    def categories(self, value):
+        self.dtype = CategoricalDtype(
+            categories=value, ordered=self.dtype.ordered,
+        )
```

Comment:
Just a note, if we're planning to expose these functions directly up to Series / Index then they should only be accessible from the CategoricalAccessor and not from the Series / Index objects directly.

Reply:
As discussed offline, we don't plan to expose Column methods as methods of Series/Index